python/sglang/srt/warmup_deepgemm.py

```diff
@@ -0,0 +1,301 @@
+
+By default, the warmup is executed during startup of SGLang. However, this means SGLang will occupy GPU memory during warmup.
+If the user wants to avoid this, it can be done by directly calling the following command without starting a SGLang server:
+
+```
+python3 -m sglang.srt.warmup_deepgemm warmup DeepseekV3ForCausalLM
+```
+
+## How to support warmup of a new model
+
+1. Capture
+(We choose to `--disable-deepgemm-warmup` here to avoid the warmup pollute the captured results)
+mkdir -p /tmp/warmup_deepgemm_capture && rm -rf /tmp/warmup_deepgemm_capture/*
+SGLANG_ENABLE_WARMUP_DEEPGEMM_CAPTURE=1 python3 -m sglang.launch_server --disable-deepgemm-warmup ...args-for-your-case...
+
+2. Analyze
+python3 -m sglang.srt.warmup_deepgemm analyze
+
+3. Update code
+According to the results of `analyze` script, update places like `_compute_sources_deepseek`
+
+## How to test the warmup
+
+1. Hack DeepGEMM's compilation code
+vim /usr/local/lib/python3.10/dist-packages/deep_gemm/jit/compiler.py
+Add prints around the line of `subprocess.check_call` (calling NVCC)
+
+2. Clear cache
+rm -rf ~/.deep_gemm
+
+3. Execute your command
+python3 -m sglang.launch_server ...args-for-your-case...
+
+4. Check whether there are any prints indicating DeepGEMM is compiling during server execution.
+"""
+
+import json
+import logging
+import multiprocessing as mp
+import sys
+from concurrent.futures import ProcessPoolExecutor
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Tuple
+
+import psutil
+import torch
+from torch.cuda import OutOfMemoryError
+from tqdm import tqdm
+
+from sglang.srt.distributed import get_tensor_model_parallel_rank
+from sglang.srt.utils import deduplicate, get_bool_env_var
+
+try:
+    import deep_gemm
+    import deep_gemm.jit_kernels.gemm
+    from deep_gemm import ceil_div, get_col_major_tma_aligned_tensor
+except ImportError:
+    pass
+
+logger = logging.getLogger(__name__)
+
+
+# --------------------------------------- warmup -------------------------------------
+
+
+def warmup(server_args, model, gpu_id):
+    from sglang.srt.layers.quantization.fp8_kernel import enable_jit_deepgemm
+
+    if (
+        server_args.disable_deepgemm_warmup
+        or (not enable_jit_deepgemm)
+        or (gpu_id != server_args.base_gpu_id)
+    ):
+        return
+
+    _warmup_raw(model.__class__.__name__)
+
+
+def _warmup_raw(model_name: str):
+    sources = _compute_sources(model_name)
+    if sources is None:
+        return
+
+    infos = _compute_infos_from_sources(sources)
+    _warmup_by_infos(infos)
+
+
+def _compute_sources(model) -> Optional[List[Dict[str, Any]]]:
+    if model in ["DeepseekV2ForCausalLM", "DeepseekV3ForCausalLM"]:
+        return _compute_sources_deepseek()
+    return None
+
+
+def _compute_sources_deepseek() -> List[Dict[str, Any]]:
+    return [
+        # TODO what is a good m_max value? (too big then we waste time warmup edge-case kernels; too small then we do not handle all cases)
```

Comment:
We should consider H100/H200/H800/H20

Reply:
`m_max` is related to the number of input tokens, so it is roughly like what context length do we want.